README.rst
===========================
Cookiecutter Django Package
===========================

.. image:: https://travis-ci.org/pydanny/cookiecutter-djangopackage.svg?branch=master
    :target: https://travis-ci.org/pydanny/cookiecutter-djangopackage

A cookiecutter_ template for creating reusable Django packages (installable apps) quickly.

**Why?** Creating reusable Django packages has always been annoying. There are no defined/maintained
best practices (especially for ``setup.py``), so you end up cutting and pasting hacky, poorly understood,
often legacy code from one project to the other. This template, inspired by `cookiecutter-pypackage`_,
is designed to allow Django developers the ability to break free from cargo-cult configuration and follow
a common pattern dictated by the experts and maintained here.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _cookiecutter-pypackage: https://github.com/audreyr/cookiecutter-pypackage

Features
--------

* Sane setup.py for easy PyPI registration/distribution
* Travis-CI configuration
* Codecov configuration
* Tox configuration
* Sphinx Documentation
* BSD licensed by default
* Basic model generation

Usage
-----

First, create your empty repo on Github (in our example below, we would call it ``blogging_for_humans``) and set up your virtual environment with your favorite method.

**Note**: Your project will be created with README.rst file containing a pypi badge, a travis-ci badge and a link to documentation on readthedocs.io. You don't need to have these accounts set up before using Cookiecutter or cookiecutter-djangopackage.

Now, get Cookiecutter_. Trust me, it's awesome::

    $ pip install cookiecutter

Now run it against this repo::

    $ cookiecutter gh:pydanny/cookiecutter-djangopackage

You'll be prompted for some questions, answer them, then it will create a directory that is your new package.

Let's pretend you want to create a reusable Django app called "Blogging-for-Humans", with an app that can be placed
in ``INSTALLED_APPS`` as "blogging_for_humans". Rather than have to copy/paste from other people's projects and
then fight enthusiasm-destroying app layout issues like `setup.py` configuration and creating test
harnesses, you get Cookiecutter_ to do all the work.

**Warning**: After this point, change 'Daniel Greenfeld', 'pydanny', etc to your own information.

It prompts you for information that it uses to create the app, with defaults in square brackets. Answer them::

    Cloning into 'cookiecutter-djangopackage'...
    remote: Counting objects: 49, done.
    remote: Compressing objects: 100% (33/33), done.
    remote: Total 49 (delta 6), reused 48 (delta 5)
    Unpacking objects: 100% (49/49), done.
    full_name [Your full name here]: Daniel Roy Greenfeld
    email [you@example.com]: pydanny@gmail.com
    github_username [yourname]: pydanny
    project_name [dj-package]: Blogging-for-Humans
    repo_name [blogging_for_humans]:
    app_name [blogging_for_humans]:
    project_short_description [Your project description goes here]: A sample Django package
    models [Comma-separated list of models]: Scoop, Flavor
    django_versions [1.11,2.1]:
    version [0.1.0]:
    Select open_source_license:
    1 - MIT
    2 - BSD
    3 - ISCL
    4 - Apache Software License 2.0
    5 - Not open source
    Choose from 1, 2, 3, 4, 5 [1]:

Enter the project and take a look around::

    $ cd blogging_for_humans/
    $ ls

Create a GitHub repo and push it there::

    $ git init
    $ git add .
    $ git commit -m "first awesome commit"
    $ git remote add origin git@github.com:pydanny/blogging_for_humans.git
    $ git push -u origin master

Now take a look at your repo. Awesome, right?

It's time to write the code!!!

Running Tests
~~~~~~~~~~~~~

Code has been written, but does it actually work? Let's find out!

::

    source <YOURVIRTUALENV>/bin/activate
    (myenv) $ pip install -r requirements_test.txt
    (myenv) $ python runtests.py

Setting up Travis
~~~~~~~~~~~~~~~~~

You will need to explicitly activate your repo in your `Travis CI profile`_.
If the repo isn't showing up, run a manual synchronisation.

.. _Travis CI profile: https://travis-ci.org/profile/

Integration with codecov.io
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Code coverage is integrated with `Codecov`_. Make sure you have an account
and that you've granted access to your repo. In case of a private repo, you
will need to generate a token and pass it when submitting coverage.

.. _CodeCov: https://codecov.io/

Register on PyPI
~~~~~~~~~~~~~~~~

Once you've got at least a prototype working and tests running, it's time to register the app on PyPI::

    python setup.py register


Releasing on PyPI
~~~~~~~~~~~~~~~~~

Time to release a new version? Easy!

First, use `bumpversion` to up the release number::

    $ pip install bumpversion
    $ bumpversion --current-version VERSION_NUMBER minor --config-file setup.cfg

Where `VERSION_NUMBER` is the current version, e.g. `0.1.0`.

Then run::

    $ python setup.py publish

It will answer with something like::

    You probably want to also tag the version now:
          git tag -a 0.1.0 -m 'version 0.1.0'
          git push --tags

Go ahead and follow those instructions.

Add to Django Packages
~~~~~~~~~~~~~~~~~~~~~~

Once you have a release, and assuming you have an account there,
just go to https://www.djangopackages.com/packages/add/ and add it there.

Follows Best Practices
~~~~~~~~~~~~~~~~~~~~~~

This project follows best practices as espoused in `Two Scoops of Django: Best Practices for Django 3.x`_.

.. _`Two Scoops of Django: Best Practices for Django 3.x`: http://twoscoopspress.org/products/two-scoops-of-django-3-x


{{cookiecutter.repo_name}}/HISTORY.rst
.. :changelog:

History
-------

{{ cookiecutter.version }} ({% now 'local' %})
__________________

* First release on PyPI.


tests/test_bake_project.py
import os
from contextlib import contextmanager

import pytest
import sh
from sh import flake8
from cookiecutter.utils import rmtree


@contextmanager
def inside_dir(dirpath):
    """
    Execute code from inside the given directory
    :param dirpath: String, path of the directory the command is being run.
    """
    old_path = os.getcwd()
    try:
        os.chdir(dirpath)
        yield
    finally:
        os.chdir(old_path)


@contextmanager
def bake_in_temp_dir(cookies, *args, **kwargs):
    """
    Delete the temporal directory that is created when executing the tests
    :param cookies: pytest_cookies.Cookies, cookie to be baked and its
    temporal files will be removed
    """
    result = cookies.bake(*args, **kwargs)
    try:
        yield result
    finally:
        rmtree(str(result.project))


def test_bake_selecting_license(cookies):
    """
    Test to check if the LICENSE gets the correct license selected
    """
    license_strings = {
        "Apache Software License 2.0": "Apache",
        "BSD": "Redistributions of source code must retain the above copyright notice, this",
        "ISCL": "Permission to use, copy, modify, and/or distribute this software for any purpose with or without fee",
        "MIT": "MIT",
    }
    for license, target_string in license_strings.items():
        with bake_in_temp_dir(
            cookies, extra_context={"open_source_license": license}
        ) as result:
            assert target_string in result.project.join("LICENSE.md").read()
            assert license in result.project.join("setup.py").read()


def test_readme(cookies):
    extra_context = {"app_name": "helloworld"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:

        readme_file = result.project.join("README.rst")
        readme_lines = [x.strip() for x in readme_file.readlines(cr=False)]
        assert "Add it to your `INSTALLED_APPS`:" in readme_lines
        assert "(myenv) $ pip install tox" in readme_lines


def test_models(cookies):
    extra_context = {"models": "ChocolateChip,Zimsterne", "app_name": "cookies"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:

        model_file = result.project.join("cookies", "models.py")
        model_txt = model_file.read()
        assert "TimeStampedModel" in model_txt

def test_views_with_models(cookies):
    """
    Test case to assert if the views are created when the models are passed
    """
    extra_context = {"models": "Pug,Dog", "app_name": "cookies"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:
        views_file = result.project.join("cookies", "views.py")
        views_file_txt = views_file.read()
        views = ["CreateView", "DeleteView", "DetailView", "UpdateView", "ListView"]
        for view in views:
            assert "Pug{}".format(view) in views_file_txt
            assert "Dog{}".format(view) in views_file_txt


def test_views_without_models(cookies):
    """
    Test case to assert that the views.py file is empty when there are no
    models defined
    """
    extra_context = {"app_name": "cookies"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:
        views_file = result.project.join("cookies", "views.py")
        views_file_txt = views_file.read()
        assert views_file_txt == ""


def test_urls_regex_with_model(cookies):
    """
    Test case to assert that the urls.py file is created when models are passed
    """
    extra_context = {"models": "Pug,Dog", "app_name": "cookies"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:
        urls_file = result.project.join("cookies", "urls.py")
        urls_file_txt = urls_file.read()
        for model in extra_context["models"].split(","):
            assert "^{}/~create/$".format(model) in urls_file_txt
            assert "^{}/(?P<pk>\d+)/~delete/$".format(model) in urls_file_txt
            assert "^{}/(?P<pk>\d+)/$".format(model) in urls_file_txt
            assert "^{}/(?P<pk>\d+)/~update/$".format(model) in urls_file_txt
            assert "^{}/$".format(model) in urls_file_txt


def test_urls_without_model(cookies):
    """
    Test case to assert that the urls.py file has the basic template when
    there are no models defined
    """
    extra_context = {"app_name": "cookies"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:
        urls_file = result.project.join("cookies", "urls.py")
        urls_file_txt = urls_file.read()
        basic_url = "url(r'', TemplateView.as_view(template_name=\"base.html\"))"  # noqa
        assert basic_url in urls_file_txt


def test_templates(cookies):
    pass


def test_tox(cookies):
    extra_context = {"app_name": "cookie_lover"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:

        tox_file = result.project.join("tox.ini")
        tox_text = tox_file.read()
        assert "commands = coverage run --source cookie_lover runtests.py" in tox_text


def test_authors(cookies):
    extra_context = {"full_name": "Cookie McCookieface"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:

        authors_file = result.project.join("AUTHORS.rst")
        authors_text = authors_file.read()
        assert "Cookie McCookieface" in authors_text



def test_manifest(cookies):
    extra_context = {"app_name": "cookie_lover"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:

        manifest_file = result.project.join("MANIFEST.in")
        manifest_text = manifest_file.read()
        assert (
            "recursive-include cookie_lover *.html *.png *.gif *js *.css *jpg *jpeg *svg *py"
            in manifest_text
        )


def test_setup_py(cookies):
    extra_context = {"app_name": "cookie_lover", "full_name": "Cookie McCookieface"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:

        setup_file = result.project.join("setup.py")
        setup_text = setup_file.read()
        assert "version=version" in setup_text
        assert "    author='Cookie McCookieface'," in setup_text


def test_django_versions_default(cookies):
    """
    Test case to assert that the tox.ini & setup.py files are generated
    with correct versions w default Django versions
    """

    with bake_in_temp_dir(cookies) as result:

        tox_file = result.project.join("tox.ini")
        tox_text = tox_file.read()

        assert "{py36,py37,py38,py39}-django-22" in tox_text
        assert "{py36,py37,py38,py39}-django-32" in tox_text
        setup_file = result.project.join("setup.py")
        setup_text = setup_file.read()
        assert "'Framework :: Django :: 2.2'," in setup_text
        assert "'Framework :: Django :: 3.2'," in setup_text
        assert "'Programming Language :: Python :: 3.6'," in setup_text
        assert "'Programming Language :: Python :: 3.7'," in setup_text
        assert "'Programming Language :: Python :: 3.8'," in setup_text
        assert "'Programming Language :: Python :: 3.9'" in setup_text


def test_new_django_versions(cookies):
    """
    Test case to assert that the tox.ini & setup.py files are generated
    with correct versions with a new Django version
    """

    extra_context = {"django_versions": "2.2,3.2", "python_versions": "py36,py37"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:

        tox_file = result.project.join("tox.ini")
        tox_text = tox_file.read()
        assert "{py36,py37}-django-22" in tox_text
        assert "{py36,py37}-django-32" in tox_text
        setup_file = result.project.join("setup.py")
        setup_text = setup_file.read()
        assert "'Framework :: Django :: 2.2'," in setup_text
        assert "'Framework :: Django :: 3.2'," in setup_text
        assert "'Programming Language :: Python :: 3.6'," in setup_text
        assert "'Programming Language :: Python :: 3.7'," in setup_text
        assert "'Programming Language :: Python :: 3.8'," not in setup_text
        assert "'Programming Language :: Python :: 3.9'," not in setup_text


# We have problems with sh and flake8 does not exists https://amoffat.github.io/sh/
# def test_flake8_compliance(cookies):
#     """generated project should pass flake8"""
#     extra_context = {"create_example_project": "Y"}
#     with bake_in_temp_dir(cookies, extra_context=extra_context) as result:
#         for file_obj in result.project.listdir():
#             name = os.path.join(file_obj.dirname, file_obj.basename)
#             if not name.endswith(".py"):
#                 continue
#             try:
#                 # sh.flake8(name)
#                 flake8(name)
#             except sh.ErrorReturnCode as e:
#                 pytest.fail(str(e))


def test_app_config(cookies):
    extra_context = {"app_name": "cookie_lover"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:

        apps_file = result.project.join("cookie_lover", "apps.py")
        apps_text = apps_file.read()
        assert "CookieLoverConfig" in apps_text
        assert "name = 'cookie_lover'" in apps_text
        readme_file = result.project.join("README.rst")
        readme_text = readme_file.read()
        assert "'cookie_lover'," in readme_text


def test_pre_commit_config(cookies):
    extra_context = {"app_name": "cookie_lover"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:
        apps_file = result.project.join(".pre-commit-config.yaml")
        apps_text = apps_file.read()
        assert "black" in apps_text
        assert "flake8" in apps_text
        assert "isort" in apps_text
        assert "mypy" in apps_text
        assert "pre-commit-hooks" in apps_text
        assert "check-merge-conflict" in apps_text
        assert "django-test" in apps_text


def test_setup_cfg(cookies):
    extra_context = {"app_name": "cookie_lover"}
    with bake_in_temp_dir(cookies, extra_context=extra_context) as result:
        apps_file = result.project.join("setup.cfg")
        apps_text = apps_file.read()
        assert "[wheel]" in apps_text
        assert "[flake8]" in apps_text
        assert "[zest.releaser]" in apps_text
        assert "[coverage:run]" in apps_text
        assert "[tool.black]" in apps_text
        assert "[metadata]" in apps_text
        assert "[mypy]" in apps_text
        assert "[tool.isort]" in apps_text
        assert "[mypy-*.migrations.*]" in apps_text
        assert "[testenv:report]" in apps_text
        assert "[testenv:reporthtml]" in apps_text
        assert "[testenv:reportxml]" in apps_text
        assert "[tox:tox]" in apps_text


# example project tests from here on


def test_make_migrations(cookies):
    """generated project should be able to generate migrations"""
    with bake_in_temp_dir(cookies, extra_context={}) as result:
        res = result.project.join("manage.py")
        try:
            sh.python(res, "makemigrations")
        except sh.ErrorReturnCode as e:
            pytest.fail(str(e))


def test_run_tests(cookies):
    """generated project should run tests"""
    with bake_in_temp_dir(cookies, extra_context={}) as result:
        res = result.project.join("runtests.py")
        try:
            with res.dirpath().as_cwd():
                sh.python(res)
        except sh.ErrorReturnCode as e:
            pytest.fail(str(e))

{{cookiecutter.repo_name}}/tests/urls.py
# -*- coding: utf-8 -*-
from __future__ import absolute_import, unicode_literals

from django.conf.urls import include, url
from django.contrib import admin

urlpatterns = [
    url(r"^admin/", admin.site.urls),
    url(
        r"^",
        include(
            "{{ cookiecutter.app_name }}.urls", namespace="{{ cookiecutter.app_name }}"
        ),
    ),
]


{{cookiecutter.repo_name}}/tests/settings.py
"""
Django settings for example project.

Generated by Cookiecutter Django Package
"""

import os

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.9/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = "{% for i in range(0, 50) %}{{ 'abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+)' | random }}{% endfor %}"

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

ALLOWED_HOSTS = []

# Application definition

INSTALLED_APPS = [
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    'django.contrib.staticfiles',

    '{{ cookiecutter.app_name }}',

    # if your app has other dependencies that need to be added to the site
    # they should be added here
]

MIDDLEWARE = [
    'django.middleware.security.SecurityMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
]

ROOT_URLCONF = 'tests.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [os.path.join(BASE_DIR, 'templates'), ],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.request',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
            ],
        },
    },
]

WSGI_APPLICATION = 'tests.wsgi.application'

# Database
# https://docs.djangoproject.com/en/1.9/ref/settings/#databases

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

# Password validation
# https://docs.djangoproject.com/en/1.9/ref/settings/#auth-password-validators

AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]

# Internationalization
# https://docs.djangoproject.com/en/1.9/topics/i18n/

LANGUAGE_CODE = 'en-us'

TIME_ZONE = 'UTC'

USE_I18N = True

USE_L10N = True

USE_TZ = True

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.9/howto/static-files/

STATIC_URL = '/static/'


{{cookiecutter.repo_name}}/{{cookiecutter.app_name}}/urls.py
# -*- coding: utf-8 -*-
from django.conf.urls import url
from django.views.generic import TemplateView

from . import views

app_name = '{{ cookiecutter.app_name }}'
urlpatterns = [
    {% if cookiecutter.models == "Comma-separated list of models" -%}
    url(r'', TemplateView.as_view(template_name="base.html")),
    {% else -%}
	{% for model in cookiecutter.models.split(',') -%}
    url(
        regex="^{{ model.strip() }}/~create/$",
        view=views.{{ model.strip() }}CreateView.as_view(),
        name='{{ model.strip() }}_create',
    ),
    url(
        regex="^{{ model.strip() }}/(?P<pk>\d+)/~delete/$",
        view=views.{{ model.strip() }}DeleteView.as_view(),
        name='{{ model.strip() }}_delete',
    ),
    url(
        regex="^{{ model.strip() }}/(?P<pk>\d+)/$",
        view=views.{{ model.strip() }}DetailView.as_view(),
        name='{{ model.strip() }}_detail',
    ),
    url(
        regex="^{{ model.strip() }}/(?P<pk>\d+)/~update/$",
        view=views.{{ model.strip() }}UpdateView.as_view(),
        name='{{ model.strip() }}_update',
    ),
    url(
        regex="^{{ model.strip() }}/$",
        view=views.{{ model.strip() }}ListView.as_view(),
        name='{{ model.strip() }}_list',
    ),
	{% endfor -%}
    {% endif -%}
]


hooks/post_gen_project.py
"""
Does the following:
# 1. Removes the example project if it isn't going to be used
"""

import os
import shutil
from pathlib import Path

PROJECT_DIRECTORY = os.path.realpath(os.path.curdir)


def remove_example_project(project_directory):
    """Removes the taskapp if celery isn't going to be used"""
    # Determine the local_setting_file_location
    location = os.path.join(PROJECT_DIRECTORY, "tests", "example")
    shutil.rmtree(location)


def insert_python_versions(versions: str = "", repo_name: str = "", app_name: str = ""):
    env_text = Path(".github/workflows/actions.yml").read_text()
    versions_arr = ""
    if "py36" in versions:
        versions_arr += "3.6, "
    if "py37" in versions:
        versions_arr += "3.7, "
    if "py38" in versions:
        versions_arr += "3.8, "
    if "py39" in versions:
        versions_arr += "3.9 "

    new_env_text = env_text.replace("__PYTHON_VERSIONS__", str(versions_arr))
    new_env_text = new_env_text.replace("__REPO_NAME__", str(repo_name))
    new_env_text = new_env_text.replace("__APP_NAME__", str(app_name))

    Path(".github/workflows/actions.yml").write_text(new_env_text)
    print("Added python versions to .github/workflows/actions.yml file.")


insert_python_versions("{{cookiecutter.python_versions}}", "{{ cookiecutter.repo_name }}", "{{ cookiecutter.app_name }}")
